lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -431,6 +431,18 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
   auto Lookup = lookupUnqualified(DC, Name, Loc, LookupOptions);
 
   if (!Lookup) {
+    // Recover from not finding 'Self' in a nominal type, if we are
+    // in a type context we can return the TypeExpr
+    if (Name.getBaseName() == Context.Id_Self) {
```

Comment:
Nitpick: this should also check that there are no argument labels (`Name.isSimpleName()`).